File: frontend/lib/merkle.ts
/**
 * Independent Merkle re-derivation, for the verification explorer and the recipient portal.
 *
 * This is the fourth implementation of the same scheme — alongside `AllocationMerkle.sol` (verifies
 * on-chain), `extension/internal/engine/merkle.go` (builds inside the enclave), and
 * `test/helpers/MerkleBuilder.sol` (builds in tests). That is deliberate rather than wasteful: a
 * recipient about to submit a claim should be able to check their proof against the on-chain root
 * *before* spending gas, using code that does not share a bug with the code that produced it.
 *
 * The conventions all four must share:
 *
 *   leaf = keccak256(keccak256(abi.encode(roundId, index, recipient, amount)))
 *
 * The outer hash keeps the leaf domain distinct from the internal-node domain, so a 64-byte
 * internal node cannot be re-presented as leaf preimage data. Internal nodes hash the *sorted*
 * pair, so a proof is a bare sibling list with no direction bits. A level with an odd node count
 * promotes the trailing node unchanged — not duplicated (which would admit a forged proof for a
 * leaf that does not exist) and not paired with a zero hash.
 */

import { concatHex, encodeAbiParameters, keccak256, type Address, type Hex } from "viem";

/** One row of an allocation table, as a recipient receives it in a disclosure. */
export interface AllocationLeaf {
  roundId: bigint;
  index: bigint;
  recipient: Address;
  amount: bigint;
}

/**
 * Computes the Merkle leaf for one allocation entry.
 *
 * `roundId` is inside the leaf so a proof cannot be replayed against another round that happens to
 * share a root. `recipient` is inside it so a leaked proof is worthless to anyone else — the claim
 * function passes `msg.sender`, never a caller-supplied address.
 */
export function leafHash({ roundId, index, recipient, amount }: AllocationLeaf): Hex {
  const inner = keccak256(
    encodeAbiParameters(
      [{ type: "uint256" }, { type: "uint256" }, { type: "address" }, { type: "uint256" }],
      [roundId, index, recipient, amount],
    ),
  );
  return keccak256(inner);
}

/** Commutative pair hash. Ordering the operands is what removes the need for direction bits. */
export function hashPair(a: Hex, b: Hex): Hex {
  return BigInt(a) < BigInt(b) ? keccak256(concatHex([a, b])) : keccak256(concatHex([b, a]));
}

/**
 * Recomputes a root from a leaf and its proof.
 *
 * A zero root means the round is not finalized. Treat that as unverifiable rather than letting an
 * empty proof against an empty root succeed.
 */
export function verifyProof(root: Hex, leaf: Hex, proof: Hex[]): boolean {
  if (BigInt(root) === 0n) return false;
  return proof.reduce<Hex>((acc, sibling) => hashPair(acc, sibling), leaf) === root;
}

/** Verifies that a recipient is entitled to an amount, against the on-chain root. */
export function verifyAllocation(root: Hex, entry: AllocationLeaf, proof: Hex[]): boolean {
  return verifyProof(root, leafHash(entry), proof);
}

/**
 * Every intermediate hash on the path from leaf to root.
 *
 * The explorer renders these so a reader can follow the derivation step by step instead of taking
 * a boolean on faith. When a proof fails, seeing where it diverges is the difference between a
 * diagnosable problem and a mystery.
 */
export interface ProofTrace {
  leaf: Hex;
  steps: Array<{ current: Hex; sibling: Hex; next: Hex; siblingIsLeft: boolean }>;
  computedRoot: Hex;
  expectedRoot: Hex;
  valid: boolean;
}

export function traceProof(root: Hex, entry: AllocationLeaf, proof: Hex[]): ProofTrace {
  const leaf = leafHash(entry);
  const steps: ProofTrace["steps"] = [];

  let current = leaf;
  for (const sibling of proof) {
    const siblingIsLeft = BigInt(sibling) < BigInt(current);
    const next = hashPair(current, sibling);
    steps.push({ current, sibling, next, siblingIsLeft });
    current = next;
  }

  return {
    leaf,
    steps,
    computedRoot: current,
    expectedRoot: root,
    valid: BigInt(root) !== 0n && current === root,
  };
}

/**
 * Builds a tree over a full allocation table.
 *
 * Only useful to someone holding the whole table — the organization, or an auditor a round has been
 * revealed to. A recipient never has this; they hold one leaf and a proof. Kept here so an
 * organization can confirm, before committing, that its policy will produce the root it expects.
 */
export function buildTree(leaves: Hex[]): { root: Hex; levels: Hex[][] } {
  if (leaves.length === 0) {
    throw new Error("cannot build a Merkle tree over zero leaves");
  }

  const levels: Hex[][] = [[...leaves]];
  let level: Hex[] = [...leaves];

  while (level.length > 1) {
    const next: Hex[] = [];
    for (let i = 0; i < level.length; i += 2) {
      const left = level[i];
      if (left === undefined) break; // unreachable: i < level.length

      const right = level[i + 1];
      // Odd trailing node: promoted unchanged, matching the Go and Solidity implementations.
      // Not duplicated — that would admit a forged proof for a leaf that does not exist.
      next.push(right === undefined ? left : hashPair(left, right));
    }
    levels.push(next);
    level = next;
  }

  const root = level[0];
  if (root === undefined) {
    throw new Error("internal: tree collapsed to an empty level");
  }
  return { root, levels };
}

/** Produces the proof for one leaf of a locally built tree. */
export function buildProof(levels: Hex[][], index: number): Hex[] {
  const proof: Hex[] = [];
  let idx = index;

  for (let depth = 0; depth < levels.length - 1; depth++) {
    const level = levels[depth];
    if (level === undefined) break;

    // A promoted trailing node has no sibling at this level and contributes nothing to the proof.
    const sibling = level[idx ^ 1];
    if (sibling !== undefined) {
      proof.push(sibling);
    }
    idx >>= 1;
  }

  return proof;
}

/** The plaintext a recipient recovers by decrypting a DISCLOSE reply. */
export interface Disclosure {
  roundId: string;
  index: number;
  recipient: Address;
  amount: string;
  proof: Hex[];
  merkleRoot: Hex;
  totalCount: number;
  engineVersion: string;
  computedAt: number;
}

/**
 * Checks a decrypted disclosure against the root the chain actually holds.
 *
 * The enclave reports the root it computed, but a recipient should not have to take that on trust —
 * this compares against the on-chain value and re-derives the proof locally, so a claim is only
 * submitted once it is known to succeed.
 */
export function checkDisclosure(
  disclosure: Disclosure,
  onChainRoot: Hex,
): { rootMatches: boolean; proofValid: boolean; ok: boolean } {
  const rootMatches = disclosure.merkleRoot.toLowerCase() === onChainRoot.toLowerCase();

  const proofValid = verifyAllocation(
    onChainRoot,
    {
      roundId: BigInt(disclosure.roundId),
      index: BigInt(disclosure.index),
      recipient: disclosure.recipient,
      amount: BigInt(disclosure.amount),
    },
    disclosure.proof,
  );

  return { rootMatches, proofValid, ok: rootMatches && proofValid };
}

File: extension/internal/engine/merkle.go
package engine

import (
	"math/big"

	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/crypto"
)

// Merkle construction, byte-compatible with contracts/libraries/AllocationMerkle.sol.
//
// Determinism here is a security property, not a performance detail: the root is the only thing
// about an allocation table that reaches the chain, and a verifier's whole argument is that
// re-running the published engine over the same policy reproduces the same root. Anything that
// varies between runs — map iteration, unstable sorts, floating point — would break that.
//
// Conventions, which must match AllocationMerkle.sol and test/helpers/MerkleBuilder.sol:
//
//   - leaf = keccak256(keccak256(abi.encode(roundId, index, recipient, amount)))
//     The outer hash keeps the leaf domain distinct from the internal-node domain, so a 64-byte
//     internal node cannot be re-presented as leaf preimage data.
//
//   - internal nodes hash the *sorted* pair, so a proof is a bare sibling list with no direction
//     bits.
//
//   - a level with an odd node count promotes the trailing node unchanged. It is not duplicated
//     (which would admit a forged proof for a non-existent leaf) and not paired with a zero hash
//     (which would make the tree shape depend on a magic constant).

// LeafHash computes the Merkle leaf for one allocation entry.
func LeafHash(roundID *big.Int, index uint64, recipient common.Address, amount *big.Int) common.Hash {
	buf := make([]byte, 0, 128)
	buf = append(buf, common.LeftPadBytes(roundID.Bytes(), 32)...)
	buf = append(buf, common.LeftPadBytes(new(big.Int).SetUint64(index).Bytes(), 32)...)
	buf = append(buf, common.LeftPadBytes(recipient.Bytes(), 32)...)
	buf = append(buf, common.LeftPadBytes(amount.Bytes(), 32)...)

	inner := crypto.Keccak256(buf)
	return common.BytesToHash(crypto.Keccak256(inner))
}

// MerkleTree holds every level of a built tree, so proofs can be produced without rebuilding.
type MerkleTree struct {
	// levels[0] is the leaves; the last level holds the single root.
	levels [][]common.Hash
}

// BuildMerkleTree constructs a tree over leaves. It returns nil for an empty input, which callers
// must treat as an error — an empty tree has no meaningful root, and returning the zero hash would
// let an empty allocation masquerade as a real one.
func BuildMerkleTree(leaves []common.Hash) *MerkleTree {
	if len(leaves) == 0 {
		return nil
	}

	// Copy: the caller's slice must not be aliased into the tree, or a later mutation would
	// silently change a root that has already been signed.
	level := make([]common.Hash, len(leaves))
	copy(level, leaves)

	levels := [][]common.Hash{level}
	for len(level) > 1 {
		level = nextLevel(level)
		levels = append(levels, level)
	}

	return &MerkleTree{levels: levels}
}

// Root returns the tree root.
func (t *MerkleTree) Root() common.Hash {
	last := t.levels[len(t.levels)-1]
	return last[0]
}

// Proof returns the sibling hashes proving the leaf at index, leaf level first.
func (t *MerkleTree) Proof(index int) []common.Hash {
	if index < 0 || index >= len(t.levels[0]) {
		return nil
	}

	proof := make([]common.Hash, 0, len(t.levels))
	idx := index
	for depth := 0; depth < len(t.levels)-1; depth++ {
		level := t.levels[depth]
		sibling := idx ^ 1
		// A promoted trailing node has no sibling at this level and contributes nothing.
		if sibling < len(level) {
			proof = append(proof, level[sibling])
		}
		idx >>= 1
	}
	return proof
}

// VerifyProof recomputes a root from a leaf and its proof. Used by the engine's own tests and by
// the ATTEST path as a self-check before signing.
func VerifyProof(root, leaf common.Hash, proof []common.Hash) bool {
	computed := leaf
	for _, sibling := range proof {
		computed = hashPair(computed, sibling)
	}
	return computed == root
}

func nextLevel(level []common.Hash) []common.Hash {
	n := (len(level) + 1) / 2
	next := make([]common.Hash, n)
	for i := 0; i < n; i++ {
		left := 2 * i
		right := left + 1
		if right < len(level) {
			next[i] = hashPair(level[left], level[right])
		} else {
			next[i] = level[left] // odd trailing node, promoted unchanged
		}
	}
	return next
}

// hashPair is commutative: the operands are ordered before hashing, so proofs need no direction
// bits and the Solidity verifier can stay branch-free.
func hashPair(a, b common.Hash) common.Hash {
	if bytesLess(a, b) {
		return common.BytesToHash(crypto.Keccak256(a.Bytes(), b.Bytes()))
	}
	return common.BytesToHash(crypto.Keccak256(b.Bytes(), a.Bytes()))
}

// bytesLess compares two hashes as big-endian unsigned integers, matching Solidity's `<` on bytes32.
func bytesLess(a, b common.Hash) bool {
	for i := 0; i < common.HashLength; i++ {
		if a[i] != b[i] {
			return a[i] < b[i]
		}
	}
	return false
}
